linkerd/core/src/main/scala/io/buoyant/linkerd/RequestEvaluator.scala

```diff
@@ -0,0 +1,165 @@
+package io.buoyant.linkerd
+
+import com.twitter.finagle._
+import com.twitter.finagle.client.Transporter.EndpointAddr
+import com.twitter.finagle.http.{MediaType, Request, Response}
+import com.twitter.finagle.naming.buoyant.DstBindingFactory
+import com.twitter.util.{Future, Return, Throw, Try}
+import io.buoyant.admin.DelegationJsonCodec
+import io.buoyant.config.Parser
+import io.buoyant.namer.DelegateTree._
+import io.buoyant.namer.{DelegateTree, Delegator}
+import io.buoyant.router.RoutingFactory
+import io.buoyant.router.RoutingFactory.BaseDtab
+import io.buoyant.router.context.{DstBoundCtx, DstPathCtx}
+
+case class EvaluatedRequest(
+  identification: String,
+  selectedAddress: String,
+  addresses: Option[Set[String]],
+  dtabResolution: List[String]
+) {
+  override def toString() = {
+    s"""
+                                 |identification: $identification
+                                 |selectedAddress: $selectedAddress
+                                 |addresses: ${addresses.getOrElse(Set.empty).mkString(",")}
+                                 |Dtab Resolution:
+                                 |${dtabResolution.mkString("\n")}
+    """.stripMargin
+  }
+}
+
+class RequestEvaluator(
+  endpoint: EndpointAddr,
+  namer: DstBindingFactory.Namer,
+  dtab: BaseDtab
+) extends SimpleFilter[Request, Response] {
+
+  private val EvaluatorHeaderName = "l5d-req-evaluate"
+  private val JsonCharSet = ";charset=UTF-8"
+
+  override def apply(
+    req: Request,
+    svc: Service[Request, Response]
+  ): Future[Response] = {
+
+    req.headerMap.get(EvaluatorHeaderName) match {
+      case None => svc(req)
+      case Some(_) =>
+
+        val identificationPath = DstPathCtx.current match {
+          case Some(dstPath) => dstPath.path
+          case None => Path.empty
+        }
+
+        val selectedEndpoint = endpoint.addr match {
+          case inetAddr: Address.Inet => inetAddr.addr.toString
+          case _ => ""
+        }
+
+        val lbSet = DstBoundCtx.current match {
+          case None => Addr.Neg
+          case Some(addrSet) => addrSet.name.addr.sample()
+        }
+
+        val addresses = lbSet match {
+          case Addr.Neg | Addr.Failed(_) | Addr.Pending => None
+          case Addr.Bound(a, _) => Some(
+            a.map { address: Address =>
```

Comment:
this can be more condensed I think:

```
a.map {
  case inetAddr: Address.Inet => ...
  case _ =>
}
```

Reply:
You're right. I think I could also have `Addr.Bound...` be the first case. I could also condense `case Addr.Neg | Addr.Failed(_) | Addr.Pending` to be the default value for `lbSet`. That way, we no longer need strict type matching used on line 67.